cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -588,7 +598,7 @@ class ObjectOutputStream : public io::OutputStream {
   }
 
   Status UploadPart(std::shared_ptr<Buffer> buffer) {
-    return UploadPart(buffer->data(), buffer->size(), buffer);
+    return UploadPart(buffer->data(), buffer->size(), std::move(buffer));
```

Comment:
Haha, it seems this crashes on MSVC because execution order of function arguments is undefined...
https://stackoverflow.com/questions/2934904/order-of-evaluation-in-c-function-parameters
@bkietz